central/virtualmachine/datastore/telemetry.go

```diff
@@ -0,0 +1,93 @@
+package datastore
+
+import (
+	"context"
+	"time"
+
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/features"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/protocompat"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/set"
+	"github.com/stackrox/rox/pkg/telemetry/phonehome"
+)
+
+const (
+	// activeAgentScanRecencyThreshold defines how recent a scan must be to consider
+	// the VM agent as active. This threshold is included in the telemetry metric name.
+	activeAgentScanRecencyThreshold = 24 * time.Hour
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Gather returns a function that collects telemetry about virtual machines.
+// It tracks three metrics:
+// 1. Number of distinct secured clusters with at least one running VM
+// 2. Total number of virtual machines
+// 3. Number of VMs with active agents (received IndexReport within last 24 hours)
+//
+// When the ROX_VIRTUAL_MACHINES feature flag is disabled, this function returns
+// an empty map without performing any database queries, ensuring no performance impact.
+func Gather(ds DataStore) phonehome.GatherFunc {
+	return func(ctx context.Context) (map[string]any, error) {
+		// Early return if virtual machines feature is disabled - zero performance impact
+		if !features.VirtualMachines.Enabled() {
+			return map[string]any{}, nil
+		}
+
+		// Use elevated permissions for telemetry gathering
+		ctx = sac.WithGlobalAccessScopeChecker(ctx,
+			sac.AllowFixedScopes(
+				sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+				sac.ResourceScopeKeys(resources.VirtualMachine),
+			),
+		)
+
+		// Fetch all virtual machines
+		vms, err := ds.SearchRawVirtualMachines(ctx, search.EmptyQuery())
```

Comment:
Does it make sense to walk the virtual machine store instead of bulk fetching everything? The issue I see is that we are not using pagination here, and it seems unnecessary to keep all VMs in memory if we just loop over them anyway.